In [21]:
import pandas as pd

# Load the dataset into a pandas DataFrame
df = pd.read_csv('/content/Dataset of Diabetes .csv')

# Display the first 5 rows of the DataFrame
print("DataFrame loaded successfully. First 5 rows:")
df.head()

DataFrame loaded successfully. First 5 rows:


ID  No_Pation Gender  AGE  Urea  Cr  HbA1c  Chol   TG  HDL  LDL  VLDL  \
0  502      17975      F   50   4.7  46    4.9   4.2  0.9  2.4  1.4   0.5   
1  735      34221      M   26   4.5  62    4.9   3.7  1.4  1.1  2.1   0.6   
2  420      47975      F   50   4.7  46    4.9   4.2  0.9  2.4  1.4   0.5   
3  680      87656      F   50   4.7  46    4.9   4.2  0.9  2.4  1.4   0.5   
4  504      34223      M   33   7.1  46    4.9   4.9  1.0  0.8  2.0   0.4   

    BMI CLASS  
0  24.0     N  
1  23.0     N  
2  24.0     N  
3  24.0     N  
4  21.0     N

In [22]:
print("Missing values per column:")
print(df.isnull().sum())

Missing values per column:
ID           0
No_Pation    0
Gender       0
AGE          0
Urea         0
Cr           0
HbA1c        0
Chol         0
TG           0
HDL          0
LDL          0
VLDL         0
BMI          0
CLASS        0
dtype: int64


In [15]:
print("Data types of 'Gender' and encoded 'CLASS' columns:")
print(df[['Gender', 'CLASS_N', 'CLASS_P', 'CLASS_Y']].dtypes)

print("\nUnique values in 'Gender' column:")
print(df['Gender'].unique())

print("\nUnique values in 'CLASS_N' column:")
print(df['CLASS_N'].unique())
print("Unique values in 'CLASS_P' column:")
print(df['CLASS_P'].unique())
print("Unique values in 'CLASS_Y' column:")
print(df['CLASS_Y'].unique())

Data types of 'Gender' and 'CLASS' columns:
Gender    object
CLASS     object
dtype: object

Unique values in 'Gender' column:
['F' 'M' 'f']

Unique values in 'CLASS' column:
['N' 'N ' 'P' 'Y' 'Y ']


In [4]:
import numpy as np

# Clean 'Gender' column: Convert 'f' to 'F'
df['Gender'] = df['Gender'].replace('f', 'F')

# Apply label encoding to 'Gender' column
df['Gender'] = df['Gender'].map({'F': 0, 'M': 1})

# Clean 'CLASS' column: Strip whitespace
df['CLASS'] = df['CLASS'].str.strip()

# Apply one-hot encoding to 'CLASS' column
df = pd.get_dummies(df, columns=['CLASS'], prefix='CLASS')

print("DataFrame after encoding 'Gender' and 'CLASS' columns:")
print(df.head())

print("\nData types after encoding:")
print(df.dtypes)

DataFrame after encoding 'Gender' and 'CLASS' columns:
    ID  No_Pation  Gender  AGE  Urea  Cr  HbA1c  Chol   TG  HDL  LDL  VLDL  \
0  502      17975       0   50   4.7  46    4.9   4.2  0.9  2.4  1.4   0.5   
1  735      34221       1   26   4.5  62    4.9   3.7  1.4  1.1  2.1   0.6   
2  420      47975       0   50   4.7  46    4.9   4.2  0.9  2.4  1.4   0.5   
3  680      87656       0   50   4.7  46    4.9   4.2  0.9  2.4  1.4   0.5   
4  504      34223       1   33   7.1  46    4.9   4.9  1.0  0.8  2.0   0.4   

    BMI  CLASS_N  CLASS_P  CLASS_Y  
0  24.0     True    False    False  
1  23.0     True    False    False  
2  24.0     True    False    False  
3  24.0     True    False    False  
4  21.0     True    False    False  

Data types after encoding:
ID             int64
No_Pation      int64
Gender         int64
AGE            int64
Urea         float64
Cr             int64
HbA1c        float64
Chol         float64
TG           float64
HDL          float64
LDL          flo

In [23]:
for col in numerical_cols:
    Q1 = df[col].quantile(0.25)
    Q3 = df[col].quantile(0.75)
    IQR = Q3 - Q1
    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR

    # Cap the outliers
    df[col] = df[col].clip(lower=lower_bound, upper=upper_bound)

print("Descriptive statistics for numerical columns after outlier capping:")
print(df[numerical_cols].describe())

Descriptive statistics for numerical columns after outlier capping:
               AGE         Urea           Cr        HbA1c         Chol  \
count  1000.000000  1000.000000  1000.000000  1000.000000  1000.000000   
mean     53.986000     4.826843    62.345000     8.280960     4.843420   
std       7.363968     1.714231    20.297906     2.532224     1.210029   
min      39.000000     0.700000    10.500000     0.950000     1.600000   
25%      51.000000     3.700000    48.000000     6.500000     4.000000   
50%      55.000000     4.600000    60.000000     8.000000     4.800000   
75%      59.000000     5.700000    73.000000    10.200000     5.600000   
max      71.000000     8.700000   110.500000    15.750000     8.000000   

                TG          HDL          LDL        VLDL          BMI  
count  1000.000000  1000.000000  1000.000000  1000.00000  1000.000000  
mean      2.280610     1.142250     2.591640     1.14040    29.566770  
std       1.150887     0.348675     1.039511     

In [24]:
print("Descriptive statistics before scaling:")
print(df[numerical_cols].describe())

Descriptive statistics before scaling:
               AGE         Urea           Cr        HbA1c         Chol  \
count  1000.000000  1000.000000  1000.000000  1000.000000  1000.000000   
mean     53.986000     4.826843    62.345000     8.280960     4.843420   
std       7.363968     1.714231    20.297906     2.532224     1.210029   
min      39.000000     0.700000    10.500000     0.950000     1.600000   
25%      51.000000     3.700000    48.000000     6.500000     4.000000   
50%      55.000000     4.600000    60.000000     8.000000     4.800000   
75%      59.000000     5.700000    73.000000    10.200000     5.600000   
max      71.000000     8.700000   110.500000    15.750000     8.000000   

                TG          HDL          LDL        VLDL          BMI  
count  1000.000000  1000.000000  1000.000000  1000.00000  1000.000000  
mean      2.280610     1.142250     2.591640     1.14040    29.566770  
std       1.150887     0.348675     1.039511     0.62744     4.926358  
min   

In [25]:
from sklearn.preprocessing import MinMaxScaler

# Initialize MinMaxScaler
min_max_scaler = MinMaxScaler()

# Apply Min-Max scaling to numerical columns
df_min_max_scaled = df.copy()
df_min_max_scaled[numerical_cols] = min_max_scaler.fit_transform(df_min_max_scaled[numerical_cols])

print("Descriptive statistics after Min-Max scaling:")
print(df_min_max_scaled[numerical_cols].describe())

Descriptive statistics after Min-Max scaling:
               AGE         Urea           Cr        HbA1c         Chol  \
count  1000.000000  1000.000000  1000.000000  1000.000000  1000.000000   
mean      0.468313     0.515855     0.518450     0.495335     0.506784   
std       0.230124     0.214279     0.202979     0.171096     0.189067   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.375000     0.375000     0.375000     0.375000     0.375000   
50%       0.500000     0.487500     0.495000     0.476351     0.500000   
75%       0.625000     0.625000     0.625000     0.625000     0.625000   
max       1.000000     1.000000     1.000000     1.000000     1.000000   

                TG          HDL          LDL         VLDL          BMI  
count  1000.000000  1000.000000  1000.000000  1000.000000  1000.000000  
mean      0.421406     0.526406     0.436503     0.400154     0.431297  
std       0.244870     0.217922     0.198002     0.241323     0.2010

In [20]:
from sklearn.preprocessing import StandardScaler

# Initialize StandardScaler
standard_scaler = StandardScaler()

# Apply Standard scaling to numerical columns
df_standard_scaled = df.copy()
df_standard_scaled[numerical_cols] = standard_scaler.fit_transform(df_standard_scaled[numerical_cols])

print("Descriptive statistics after Standard scaling:")
print(df_standard_scaled[numerical_cols].describe())

Descriptive statistics after Standard scaling:
                AGE          Urea            Cr         HbA1c          Chol  \
count  1.000000e+03  1.000000e+03  1.000000e+03  1.000000e+03  1.000000e+03   
mean   3.694822e-16 -1.847411e-16  5.151435e-17 -1.136868e-16 -5.684342e-17   
std    1.000500e+00  1.000500e+00  1.000500e+00  1.000500e+00  1.000500e+00   
min   -2.036062e+00 -2.408607e+00 -2.555482e+00 -2.896516e+00 -2.681788e+00   
25%   -4.056908e-01 -6.576751e-01 -7.070768e-01 -7.036703e-01 -6.973731e-01   
50%    1.377664e-01 -1.323955e-01 -1.155870e-01 -1.110094e-01 -3.590138e-02   
75%    6.812236e-01  5.096128e-01  5.251937e-01  7.582267e-01  6.255704e-01   
max    2.311595e+00  2.260545e+00  2.373599e+00  2.951072e+00  2.609986e+00   

                 TG           HDL           LDL          VLDL           BMI  
count  1.000000e+03  1.000000e+03  1.000000e+03  1.000000e+03  1.000000e+03  
mean  -2.131628e-16  1.030287e-16  4.618528e-17 -8.526513e-17 -3.979039e-16  
std    

In [53]:
import pandas as pd

# Load the dataset into a pandas DataFrame
df = pd.read_csv('/content/adult.csv')

# Display the first 5 rows of the DataFrame
print("DataFrame loaded successfully. First 5 rows:")
df.head()


ParserError: Error tokenizing data. C error: Expected 15 fields in line 28856, saw 20


In [56]:
print("Missing values per column:")
print(df.isnull().sum())
print()

print("Data types of the DataFrame columns:")
print(df.dtypes)
print()

print("\nUnique values for object columns (potential categorical features):")
for col in df.select_dtypes(include='object').columns:
    print(f"{col}: {df[col].nunique()} unique values")

Missing values per column:
age               0
workclass         0
fnlwgt            0
education         0
education_num     0
marital_status    0
occupation        0
relationship      0
race              0
sex               0
capital_gain      1
capital_loss      2
hours_per_week    1
native_country    0
income            0
dtype: int64

Data types of the DataFrame columns:
age                 int64
workclass          object
fnlwgt              int64
education          object
education_num       int64
marital_status     object
occupation         object
relationship       object
race               object
sex                object
capital_gain      float64
capital_loss      float64
hours_per_week    float64
native_country     object
income             object
dtype: object


Unique values for object columns (potential categorical features):
workclass: 8 unique values
education: 16 unique values
marital_status: 8 unique values
occupation: 16 unique values
relationship: 7 unique values
rac

**Reasoning**:
The previous step identified all columns as object type, including those that are inherently numerical. Before applying one-hot encoding to the specified categorical columns, I need to first convert the numerical columns ('age', 'fnlwgt', 'education_num', 'capital_gain', 'capital_loss', 'hours_per_week') to numeric types (e.g., integer or float). This ensures proper data handling and prevents issues during encoding or subsequent analysis, as one-hot encoding is not suitable for these columns. After converting numeric columns, I will proceed with one-hot encoding for the specified categorical columns, dropping the original columns to avoid multicollinearity as requested.



In [48]:
import pandas as pd

# List of columns that should be numerical but are currently 'object'
numerical_cols_to_convert = ['age', 'fnlwgt', 'education_num', 'capital_gain', 'capital_loss', 'hours_per_week']

# Convert these columns to numeric, coercing errors to NaN if any
for col in numerical_cols_to_convert:
    df[col] = pd.to_numeric(df[col], errors='coerce')

# List of columns to one-hot encode as per the instructions
categorical_cols_to_encode = [
    'workclass', 'education', 'marital_status', 'occupation',
    'relationship', 'race', 'native_country', 'sex', 'income'
]

# Apply one-hot encoding to the specified categorical columns
df_encoded = pd.get_dummies(df, columns=categorical_cols_to_encode, drop_first=False)

print("DataFrame head after converting numerical columns and one-hot encoding categorical columns:")
print(df_encoded.head())

print("\nData types after encoding:")
print(df_encoded.dtypes)


DataFrame head after converting numerical columns and one-hot encoding categorical columns:
   age  fnlwgt  education_num  capital_gain  capital_loss  hours_per_week  \
1   25  226802              7           0.0           0.0            40.0   
2   38   89814              9           0.0           0.0            50.0   
3   28  336951             12           0.0           0.0            40.0   
4   44  160323             10        7688.0           0.0            40.0   
5   18  103497             10           0.0           0.0            30.0   

   workclass_Federal-gov  workclass_Local-gov  workclass_Never-worked  \
1                  False                False                   False   
2                  False                False                   False   
3                  False                 True                   False   
4                  False                False                   False   
5                  False                False                   False   

   wor

## Handle Outliers

### Subtask:
Detect and manage outliers in the numerical features of the dataset using methods such as the IQR method or z-score, to ensure robust model performance.


**Reasoning**:
To detect and manage outliers in the specified numerical columns using the IQR method, I will iterate through each column, calculate Q1, Q3, and IQR, define the lower and upper bounds, and then cap the outliers. Finally, I will display the descriptive statistics for these columns to show the effect of the treatment.



In [49]:
import numpy as np

# Identify the numerical columns for outlier treatment
numerical_cols_for_outliers = ['age', 'fnlwgt', 'education_num', 'capital_gain', 'capital_loss', 'hours_per_week']

# Apply IQR-based outlier capping for each numerical column
for col in numerical_cols_for_outliers:
    Q1 = df_encoded[col].quantile(0.25)
    Q3 = df_encoded[col].quantile(0.75)
    IQR = Q3 - Q1

    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR

    # Cap the outliers
    df_encoded[col] = df_encoded[col].clip(lower=lower_bound, upper=upper_bound)

print("Descriptive statistics for numerical columns after IQR-based outlier capping:")
print(df_encoded[numerical_cols_for_outliers].describe())

Descriptive statistics for numerical columns after IQR-based outlier capping:
               age         fnlwgt  education_num  capital_gain  capital_loss  \
count  38463.00000   38463.000000   38463.000000       38462.0       38461.0   
mean      38.64098  186770.662988      10.121194           0.0           0.0   
std       13.66766   95202.783991       2.448241           0.0           0.0   
min       17.00000   12285.000000       4.500000           0.0           0.0   
25%       28.00000  117682.000000       9.000000           0.0           0.0   
50%       37.00000  178142.000000      10.000000           0.0           0.0   
75%       48.00000  237735.000000      12.000000           0.0           0.0   
max       78.00000  417814.500000      16.000000           0.0           0.0   

       hours_per_week  
count    38462.000000  
mean        41.180308  
std          6.180481  
min         32.500000  
25%         40.000000  
50%         40.000000  
75%         45.000000  
max      

## Apply Min-Max Scaler

### Subtask:
Apply Min-Max Scaling (Normalization) to numerical features to scale them to a fixed range, typically 0 to 1.


**Reasoning**:
To apply Min-Max Scaling, I will import `MinMaxScaler`, instantiate it, and then use `fit_transform` on the specified numerical columns within `df_encoded`. Finally, I will print the descriptive statistics of these columns to confirm the scaling.



In [50]:
from sklearn.preprocessing import MinMaxScaler

# Instantiate MinMaxScaler
min_max_scaler = MinMaxScaler()

# Apply Min-Max scaling to the specified numerical columns in df_encoded
df_encoded[numerical_cols_for_outliers] = min_max_scaler.fit_transform(df_encoded[numerical_cols_for_outliers])

print("Descriptive statistics for numerical columns after Min-Max scaling:")
print(df_encoded[numerical_cols_for_outliers].describe())

Descriptive statistics for numerical columns after Min-Max scaling:
                age        fnlwgt  education_num  capital_gain  capital_loss  \
count  38463.000000  38463.000000   38463.000000       38462.0       38461.0   
mean       0.354770      0.430266       0.488800           0.0           0.0   
std        0.224060      0.234762       0.212891           0.0           0.0   
min        0.000000      0.000000       0.000000           0.0           0.0   
25%        0.180328      0.259900       0.391304           0.0           0.0   
50%        0.327869      0.408989       0.478261           0.0           0.0   
75%        0.508197      0.555940       0.652174           0.0           0.0   
max        1.000000      1.000000       1.000000           0.0           0.0   

       hours_per_week  
count    38462.000000  
mean         0.434015  
std          0.309024  
min          0.000000  
25%          0.375000  
50%          0.375000  
75%          0.625000  
max          1.0000

## Apply Standard Scaler

### Subtask:
Apply Standard Scaling to numerical features to transform them to have a mean of 0 and a standard deviation of 1.


**Reasoning**:
To apply Standard Scaling, I will import `StandardScaler`, instantiate it, and then use `fit_transform` on the specified numerical columns within `df_encoded`. Finally, I will print the descriptive statistics of these columns to confirm the scaling.



In [38]:
from sklearn.preprocessing import StandardScaler

# Instantiate StandardScaler
standard_scaler = StandardScaler()

# Apply Standard scaling to the specified numerical columns in df_encoded
df_encoded[numerical_cols_for_outliers] = standard_scaler.fit_transform(df_encoded[numerical_cols_for_outliers])

print("Descriptive statistics for numerical columns after Standard scaling:")
print(df_encoded[numerical_cols_for_outliers].describe())

Descriptive statistics for numerical columns after Standard scaling:
                age        fnlwgt  education_num  capital_gain  capital_loss  \
count  3.846300e+04  3.846300e+04   3.846300e+04       38462.0       38461.0   
mean   9.458385e-17  3.343687e-17   2.301787e-16           0.0           0.0   
std    1.000013e+00  1.000013e+00   1.000013e+00           0.0           0.0   
min   -1.583392e+00 -1.832803e+00  -2.296043e+00           0.0           0.0   
25%   -7.785618e-01 -7.257095e-01  -4.579651e-01           0.0           0.0   
50%   -1.200645e-01 -9.063574e-02  -4.950328e-02           0.0           0.0   
75%    6.847655e-01  5.353310e-01   7.674203e-01           0.0           0.0   
max    2.879756e+00  2.426892e+00   2.401268e+00           0.0           0.0   

       hours_per_week  
count    3.846200e+04  
mean    -7.315660e-17  
std      1.000013e+00  
min     -1.404490e+00  
25%     -1.909759e-01  
50%     -1.909759e-01  
75%      6.180331e-01  
max      1.831547e

## Final Task

### Subtask:
Summarize all the data preprocessing steps performed and display the head of the processed dataset to show the final state.


## Summary:

### Data Analysis Key Findings

The data preprocessing involved a series of steps to prepare the dataset for analysis:

*   **Data Loading**: The `adult.csv` dataset was loaded into a DataFrame. Initially, a `ParserError` was encountered due to inconsistent field counts. This was resolved by using `on_bad_lines='skip'` and then dropping the first row, which contained the original header loaded as data.
*   **Missing Value Handling**:
    *   All '?' characters were replaced with `np.nan`. This revealed missing values in 'workclass' (2215), 'occupation' (2222), and 'native_country' (670), among others.
    *   All missing values were then imputed using the mode of their respective columns, resulting in a dataset with no missing values, and the row count remained at 38463.
*   **Categorical Data Encoding**:
    *   Numerical columns initially loaded as 'object' type ('age', 'fnlwgt', 'education\_num', 'capital\_gain', 'capital\_loss', 'hours\_per\_week') were converted to numeric types.
    *   Categorical features ('workclass', 'education', 'marital\_status', 'occupation', 'relationship', 'race', 'native\_country', 'sex', 'income') were one-hot encoded, expanding the DataFrame with new binary columns.
*   **Outlier Treatment**:
    *   Outliers in numerical columns ('age', 'fnlwgt', 'education\_num', 'capital\_gain', 'capital\_loss', 'hours\_per\_week') were capped using the Interquartile Range (IQR) method.
    *   For 'age', values were capped between 17.0 and 78.0. For 'hours\_per\_week', values were capped between 32.5 and 52.5.
    *   Notably, 'capital\_gain' and 'capital\_loss' had nearly all non-zero values capped to 0.0, indicating a highly skewed distribution where most non-zero values were treated as outliers.
*   **Min-Max Scaling**: The outlier-treated numerical columns were scaled to a range of \[0, 1] using `MinMaxScaler`, ensuring all values are within this normalized range.
*   **Standard Scaling**: Finally, these numerical columns were scaled using `StandardScaler` to have a mean very close to 0 and a standard deviation very close to 1. This step resulted in 'capital\_gain' and 'capital\_loss' having a standard deviation of 0.0, further indicating their predominantly zero values after previous processing.

### Insights or Next Steps

*   The extensive preprocessing addresses common data quality issues, making the dataset suitable for machine learning models that are sensitive to scale, distribution, and categorical features.
*   Given that 'capital\_gain' and 'capital\_loss' became predominantly zero after outlier treatment and scaling, further investigation into these features' information content or alternative transformation methods might be beneficial before model training, as their current state suggests limited variance.
